README.md
# Nerves.SSDPServer

**Simple Server for the <u>S</u>imple <u>S</u>ervices <u>D</u>iscovery <u>P</u>rotocol**

[SSDP](https://en.wikipedia.org/wiki/Simple_Service_Discovery_Protocol) allows devices on LAN to announce themselves and their services to other devices. 

## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed as:

  1. Add nerves_ssdp_server to your list of dependencies in `mix.exs`:

        def deps do
          [{:nerves_ssdp_server, "~> 0.2.2"}]
        end

  2. Ensure nerves_ssdp_server is started before your application:

        def application do
          [applications: [:nerves_ssdp_server]]
        end

## Usage

In SSDP, every service needs to define a USN (unique service name), and a ST (service type).  That's the minimum required to publish a service.  With that info, it's as simple as this:

```elixir
alias Nerves.SSDPServer

SSDPServer.publish "my_unique_service_name", "my-service-type"
```
### Publishing custom fields

Other parameters you might specify for the second parameter to `publish` are included as fields of the published service.  For instance, you can do..

```elixir
alias Nerves.SSDPServer

@ssdp_fields [
    location: "http://localhost:3000/myservice.json",
    server: "MyServerName",
    "cache-control": "max-age=1800"
]

SSDPServer.publish "my-service-name", "my-service-type", @ssdp_fields
```

You can call `publish` to publish multiple services, each with a unique USN, but you can only publish each USN once, as per the SSDP spec.

## The nerves-project-org:service:cell:1 service type

- LOCATION if present, specifies a URL to grab a cell description

## References

Here is are some links with background information about SSDP.

http://www.w3.org/TR/discovery-api/#simple-service-discovery-protocol-ssdp
http://www.upnp.org/specs/arch/UPnP-arch-DeviceArchitecture-v1.0-20080424.pdf



CHANGELOG.md
# Changelog

## 0.2.2 (2017-Feb-23)

* Support Elixir ~> 1.4
* Require Keyword List configuration

## 0.2.1 (2016-Jul-19)

* Cleaned up documentation
* Publish now correctly returns {:ok, usn} rather than {:ok, pid}
* Provided Hex package configuration
* Changed to Apache2 License
* Support Elixir ~> 1.2

## 0.2.0-dev (Mar 8, 2016)

* API changed to accomodate sending both USN and ST to publish
* Simplified and fixed bugs in server and message generator


test/test_helper.exs
E}xUnit.start()


test/ssdp_server_test.exs
defmodule Nerves.SSDPServerTest do

  alias Nerves.SSDPServer
  use ExUnit.Case, seed: 0

  doctest SSDPServer

  alias Nerves.SSDPClient

  @test_usn_1 "uid:test:my_usn_1"
  @test_st_1 "nerves-project-org:test-service:1"
  @test_usn_2 "uid:test:my_usn_2"
  @test_st_2 "test-nerves-project-org:another-test-service:1"

  @test_fields_1 [
    location: "http://there/",
    server: "SSDPServerTest",
    "cache-control": "max-age=1800"
  ]

  @test_fields_2 [
    location: "http://there/",
    server: "SSDPServerTest2",
    "cache-control": "max-age=5"
  ]

  test "ssdp publishing, un-publishing works" do
    # test publishing and behavior
    publish_and_test(@test_usn_1, @test_st_1, @test_fields_1)
    publish_and_test(@test_usn_2, @test_st_2, @test_fields_2)
    unpublish_and_test @test_usn_1
  end

  defp publish_and_test(usn, st, fields) do
    {:ok, ^usn} = SSDPServer.publish usn, st, fields
    responses = SSDPClient.discover
    response = responses[usn]
    assert is_map(response)
    Enum.each fields, fn({k,v}) ->
      assert response[k] == v
    end
  end

  defp unpublish_and_test(usn) do
    :ok = SSDPServer.unpublish usn
    :timer.sleep 1000
    responses = SSDPClient.discover
    assert responses[usn] == nil
  end
end


test/nerves_ssdp_server_test.exs
defmodule Nerves.SSDPServerTest do
  use ExUnit.Case
  doctest Nerves.SSDPServer

  test "the truth" do
    assert 1 + 1 == 2
  end
end


mix.exs
defmodule Nerves.SSDPServer.Mixfile do
  use Mix.Project

  @version "0.2.2"

  def project do
    [ app: :nerves_ssdp_server,
      version: @version,
      elixir: "~> 1.4",
      build_embedded: Mix.env == :prod,
      start_permanent: Mix.env == :prod,
      deps: deps(),
      description: "Client for Simple Service Discovery Protocol",
      package: package(),
      name: "Nerves.SSDPServer",
      docs: docs() ]
  end

  def application do
    [ applications: [:logger],
      mod: {Nerves.SSDPServer, []} ]
  end

  defp deps do
    [ {:nerves_ssdp_client, github: "nerves-project/nerves_ssdp_client", only: :test},
      {:ex_doc, "~> 0.11", only: :dev} ]
  end

  defp docs do
    [ source_ref: "v#{@version}",
      main: "Nerves.SSDPServer",
      source_url: "https://github.com/nerves-project/nerves_ssdp_server",
      extras: ["README.md", "CHANGELOG.md"] ]
  end

  defp package do
    [ maintainers: ["Garth Hitchens"],
      licenses: ["Apache-2.0"],
      links: %{github: "https://github.com/nerves-project/nerves_ssdp_server"},
      files: ~w(lib config) ++ ~w(README.md CHANGELOG.md LICENSE mix.exs) ]
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure for your application as:
#
#     config :nerves_ssdp_server, key: :value
#
# And access this configuration in your application as:
#
#     Application.get_env(:nerves_ssdp_server, :key)
#
# Or configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/ssdp_server/server.ex
defmodule Nerves.SSDPServer.Server do

  alias Nerves.SSDPServer.Messages
  use GenServer

  @type state :: map

  @moduledoc false

  @initial_state %{
    st: nil,
    recv_socket: nil,
    xmit_socket: nil,
    usn: nil,
    notify_count: 0,
    fields: []
  }

  def start_link(st, usn, fields) do
    GenServer.start_link(__MODULE__, [st, usn, fields], [])
  end

  # public genserver handlers

  def init([st, usn, fields]) do
    @initial_state
    |> Map.merge(%{st: st, usn: usn, fields: fields})
    |> open_ssdp_sockets!
    |> notify!
    |> reschedule_notify!
    |> tuple_reply(:ok)
  end

  # handle received udp m-search packets and ignore all others

  @msearch "M-SEARCH * HTTP/1.1"

  def handle_info({:udp, _s, ip, port, <<@msearch, rest :: binary>>}, state) do
    state
    |> handle_msearch(ip, port, rest)
    |> tuple_reply(:noreply)
  end

  def handle_info({:udp, _s, _ip, _port, _}, state), do: {:noreply, state}

  # timer handlers for notify and m-search response

  def handle_info(:notify_timer, state) do
    state
    |> notify!
    |> reschedule_notify!
    |> tuple_reply(:noreply)
  end

  def handle_info({:respond_timer, ip, port}, state) do
    state
    |> respond!(ip, port)
    |> tuple_reply(:noreply)
  end

  # ssdp m-search & reply handling (private)

  @default_search_target "ssdp:all"
  defp handle_msearch(state, ip, port, packet) do
    headers = packet |> parse_httpu_headers
    search_target = headers[:st] || @default_search_target
    if m_search_matches?(search_target, state.st) do
      headers
      |> response_time
      |> :erlang.send_after(Kernel.self, {:respond_timer, ip, port})
    end
    state
  end

  @default_mx 3000
  @spec response_time(Keyword.t) :: integer
  # return random time in milliseconds (0-MX)
  defp response_time(headers) do
    if (headers[:mx]) do
      String.to_integer(headers[:mx]) * 1000
    else
      @default_mx
    end
    |> :rand.uniform
  end

  defp respond!(state, ip, port) do
    message = Messages.response(state.usn, state.st, state.fields)
    :ok = :gen_udp.send(state.xmit_socket, ip, port, message)
    state
  end

  # search match definitions

  def m_search_matches?("ssdp:all", _st), do: true
  def m_search_matches?(target, st), do:  String.equivalent?(target, st)

  # ssdp notification (private)

  @slow_notify_after 5                # after 5 notifies slow downd
  @fast_notify_interval  3000         # every 3 seconds at first
  @slow_notify_interval  30000        # 30 seconds thereafter

  @spec notify!(state) :: state
  defp notify!(state) do
    Messages.alive(state.usn, state.st, state.fields)
    |> send_multicast_ssdp_message!(state.xmit_socket)
    %{state | notify_count: state.notify_count + 1}
  end

  @spec reschedule_notify!(state) :: state
  defp reschedule_notify!(state) do
    time_to_next_notify = notify_interval(state.notify_count)
    :erlang.send_after time_to_next_notify, Kernel.self, :notify_timer
    state
  end

  defp notify_interval(notify_count) do
    if (notify_count > @slow_notify_after) do
      @slow_notify_interval
    else
      @fast_notify_interval
    end
  end

  # socket management (private)

  @mcast_group {239,255,255,250}
  @mcast_port 1900

  defp open_ssdp_sockets!(state) do
    {:ok, recv_socket} = :gen_udp.open @mcast_port, recv_socket_opts(state)
    {:ok, xmit_socket} = :gen_udp.open 0, xmit_socket_opts(state)
    %{state | recv_socket: recv_socket, xmit_socket: xmit_socket}
  end

  defp recv_socket_opts(state), do: [
    ip: @mcast_group,
    active: true,
    mode: :binary,
    reuseaddr: true,
    multicast_loop: true,
    add_membership: {@mcast_group, multicast_if(state)},
    #multicast_if: multicast_if(state),
    #broadcast: true,
  ]

  defp xmit_socket_opts(_state), do: [
    reuseaddr: true,
    mode: :binary,
    multicast_loop: true,
    multicast_ttl: 4
    #multicast_if: multicast_if(state),
  ]

  defp multicast_if(_state) do
    {0,0,0,0}
  end

  defp send_multicast_ssdp_message!(message, socket) do
    :ok = :gen_udp.send(socket, @mcast_group, @mcast_port, message)
  end

  # transform state into a reply for use with common erlang and genserver responses
  defp tuple_reply(state, atom) when is_atom(atom), do: {atom, state}

  defp parse_httpu_headers(packet) do
   raw_params = String.split(packet, ["\r\n", "\n"])
   mapped_params = Enum.map raw_params, fn(x) ->
     case String.split(x, ":", parts: 2) do
       [k, v] -> {String.to_atom(String.downcase(k)), String.strip(v)}
       _ -> nil
     end
   end
   Enum.reject mapped_params, &(&1 == nil)
 end

end


lib/ssdp_server/messages.ex
defmodule Nerves.SSDPServer.Messages do

  @moduledoc false
  @notify_header "NOTIFY * HTTP/1.1"
  @response_header "HTTP/1.1 200 OK"

  def alive(usn, st, fields) do
    fields
    |> Keyword.merge(nt: st, usn: usn, nts: "ssdp:alive")
    |> format_response_with_header(@notify_header)
  end

  def byebye(usn, st, fields) do
    fields
    |> Keyword.merge(nt: st, usn: usn, nts: "ssdp:byebye")
    |> format_response_with_header(@notify_header)
  end

  def response(usn, st, fields) do
    fields
    |> Keyword.merge(st: st, usn: usn)
    |> format_response_with_header(@response_header)
  end

  # private

  defp format_response_with_header(fields, header) do
    header <> "\r\n" <> format_fields_for_response(fields) <> "\r\n"
  end

  defp format_fields_for_response(fields) do
    Enum.map_join fields, &transform_field_into_response_line/1
  end

  defp transform_field_into_response_line({k, v}) do
    formatted_key(k) <> ": " <> formatted_value(v) <> "\r\n"
  end

  defp formatted_key(key) do
    key
    |> :erlang.atom_to_binary(:utf8)
    |> String.upcase
  end

  defp formatted_value(v) when is_function(v), do: v.()
  defp formatted_value(v), do: to_string(v)

end


lib/ssdp_server.ex
defmodule Nerves.SSDPServer do

  @moduledoc """
  Implements a simple subset of the [Simple Service Discovery Protocol](https://en.wikipedia.org/wiki/Simple_Service_Discovery_Protocol).

  This does *not* implement the full UPNP specification, but uses the
  multicast SSDP protocol in order to provide LAN presence annoucment
  and device discovery.
  """

  @typedoc """
  Unique Service Identifier -- Uniquely identifies the service, and must be unique on the local network.
  """
  @type usn :: String.t

  @typedoc """
  Service Type -- a string that identifies the type of SSDP service.
  """

  @type st :: String.t
  use Application

  alias Nerves.SSDPServer
  alias SSDPServer.Server
  import Supervisor.Spec, warn: false

  @sup Nerves.SSDPServer.Supervisor

  @doc false
  def start(_type, _args) do
    Supervisor.start_link [], strategy: :one_for_one, name: @sup
  end

  @doc """
  Publish the service, returning the USN (unique service name) for
  the service, which can later be used to unpublish the service.

  - `usn` (unique service name) - uniquely identifies the service, and must be
    unique on the local network. attempts to publish two services with the same
    usn on the same node will result in an error.

  - `st` (service type) :: SSDP Service Type - a string that identifies the
    type of service.

  - `fields` - Keyword list consiting of fields ot be added to the SSDP replies.

  ## Examples

  ### Simple Publishing

      Nerves.SSDPServer.publish "my_unique_service_name", "my-service-type"

  ### Publishing custom fields

  Other parameters you might specify for the second paramter do `publish` are
  included as fields of the published service. For instance, you can do..

      @ssdp_fields [
          location: "http://localhost:3000/myservice.json",
          server: "MyServerName",
          "cache-control": "max-age=1800"
      ]

      Nerves.SSDPServer.publish "my-service-name", "my-service-type", @ssdp_fields
  """
  @spec publish(usn, st, Keyword.t) :: {:ok, usn} | {:error, atom}
  def publish(usn, st, fields \\ []) do
    ssdp_worker = worker(Server, [
      (st |> to_string), (usn |> to_string), fields],
      id: usn, restart: :transient)
    Supervisor.start_child(@sup, ssdp_worker)
    |> case do
      {:ok, _pid} -> {:ok, usn}
      other -> other
    end
  end

  @doc """
  Stop publishing the service designated by `usn`.
  """
  @spec unpublish(usn) :: :ok | {:error, atom}
  def unpublish(usn) do
    Supervisor.terminate_child @sup, usn
  end

end
